src/dft/src/stitch/ScanStitch.cpp

```diff
@@ -0,0 +1,182 @@
+
+  ScanDriver scan_in_port = FindOrCreateScanIn(block);
+
+  // We need fast pop for front and back
+  std::deque<std::shared_ptr<ScanCell>> scan_cells;
+
+  const std::vector<std::shared_ptr<ScanCell>>& falling_edge
+      = scan_chain.getFallingEdgeScanCells();
+  const std::vector<std::shared_ptr<ScanCell>>& rising_edge
+      = scan_chain.getRisingEdgeScanCells();
+
+  // Falling edge first
+  std::copy(
+      falling_edge.begin(), falling_edge.end(), std::back_inserter(scan_cells));
+  std::copy(
+      rising_edge.begin(), rising_edge.end(), std::back_inserter(scan_cells));
+
+  // All the cells in the scan chain are controlled by the same scan enable
+  for (const auto& scan_cell : scan_cells) {
+    scan_cell->connectScanEnable(*scan_enable_);
+  }
+
+  // Lets get the first and last cell
+  const std::shared_ptr<ScanCell>& first_scan_cell = *scan_cells.begin();
+  const std::shared_ptr<ScanCell>& last_scan_cell = *(scan_cells.end() - 1);
+
+  if (!scan_cells.empty()) {
+    scan_cells.pop_front();
+  }
+
+  if (!scan_cells.empty()) {
+    scan_cells.pop_back();
+  }
+
+  for (auto current = scan_cells.begin(); current != scan_cells.end();
+       ++current) {
+    auto next = current + 1;
+    // Connects current cell scan out to next cell scan in
+    (*next)->connectScanIn((*current)->getScanOut());
+  }
+
+  // Let's connect the first cell
+  first_scan_cell->connectScanEnable(*scan_enable_);
+  first_scan_cell->connectScanIn(scan_in_port);
+
+  if (!scan_cells.empty()) {
+    (*scan_cells.begin())->connectScanIn(first_scan_cell->getScanOut());
+  } else {
+    // If last_scan_cell == first_scan_cell, then scan in was already connected
+    if (last_scan_cell != first_scan_cell) {
+      last_scan_cell->connectScanIn(first_scan_cell->getScanOut());
+    }
+  }
+
+  // Let's connect the last cell
+  ScanLoad scan_out_port
+      = FindOrCreateScanOut(block, last_scan_cell->getScanOut());
+  last_scan_cell->connectScanOut(scan_out_port);
+}
+
+void ScanStitch::FindOrCreateScanEnable(odb::dbBlock* block)
+{
+  // TODO: For now we will create a new scan_enable pin at the top level. We
+  // need to support defining DFT signals for scan_enable
+  for (uint64_t scan_enable_number = 1;; ++scan_enable_number) {
+    std::string scan_enable_pin_name
+        = fmt::format(kScanEnableNamePattern, scan_enable_number);
+    odb::dbBTerm* scan_enable = block->findBTerm(scan_enable_pin_name.c_str());
+    if (!scan_enable) {
+      odb::dbNet* net = odb::dbNet::create(block, scan_enable_pin_name.c_str());
+      net->setSigType(odb::dbSigType::SCAN);
+      odb::dbBTerm* scan_enable_port
+          = odb::dbBTerm::create(net, scan_enable_pin_name.c_str());
+      scan_enable_port->setSigType(odb::dbSigType::SCAN);
+      scan_enable_port->setIoType(odb::dbIoType::INPUT);
+      scan_enable_.emplace(scan_enable_port);
+      return;
+    }
+  }
+}
+
+ScanDriver ScanStitch::FindOrCreateScanIn(odb::dbBlock* block)
```

Comment:
FindOrCreateScanEnable, FindOrCreateScanIn, and FindOrCreateScanOut look to share a lot code that could be factored into a common method.

Reply:
Done. I factored out the part of the creation into a private method in ScanStitch.